Release: v1.0

sgammon opened the pull request:
Initial release of Bloombox for Java. Full test coverage and a production-ready library are the goals.

Worklog:
- [x] Version bump -> `1.0`
- [x] Snapshot bump -> `1.1-SNAPSHOT`
- [x] Redundant static analysis
- [x] Full test coverage, or close to it
- [x] Test sample projects with Gradle and Maven
- [x] Flawless docs and site, or close to it
- [x] Make sure source and docs JARs work
- [x] Find and fix any straggler versions
- [x] Remove any reference to beta, release candidate, and so on
- [x] Mark docs as `production`

Enclosed work:
- [x] Feature: mTLS (bloombox/Java#22)

_UPDATE:_ Released as ~~`v1.0-rc1`~~ -> `rc9`. After testing will approve, merge, tag and release as `1.0`, assuming we don't need additional revisions.

Commit:
Release: v1.0

Initial release of Bloombox for Java. Full test coverage and a
production-ready library are the goals.

Worklog:
- [x] Version bump -> `1.0`
- [x] Snapshot bump -> `1.1-SNAPSHOT`
- [x] Redundant static analysis
- [ ] Flawless docs and site, or close to it
- [ ] Make sure source and docs JARs work
- [ ] Test sample projects with Gradle and Maven
- [ ] Full test coverage, or close to it

Commit:
Update tests to close connections immediately afterwards

Commit:
Switch to CodeClimate coverage badge

Commit:
Cache Maven home

tyler-porras commented on src/test/kotlin/bloombox/client/shop/ShopInfoTest.kt:
```diff
@@ -112,73 +116,81 @@ class ShopInfoTest: ClientRPCTest() {
-    assertEquals(responseOne.deliveryMinimum, (50.0).toFloat(), "response should specify correct delivery minimum for zipcheck 1")
-    assertEquals(responseTwo.deliveryMinimum, (200.0).toFloat(), "response should specify correct delivery minimum for zipcheck 2")
+    withClient({ client ->
+      val responseOne = client.platform.shop().checkZipcode("95120", ShopClient.ShopContext(
```
Are these Zip codes hard coded into the test to match against preset zip codes in the code base, or are we testing against Caliva's data specifically for now?

tyler-porras replied:
ie. what is a "known good zip code" in relation to these tests? A test club? Caliva's real data? etc.

sgammon replied:
we talked about this offline, but updating here for posterity's sake. yes, these are hard coded to work with the test partner/location pair and API key.

Commit:
Use production endpoints for internal API traffic, for now

Commit:
Feature: mTLS

Introduce functionality to allow mTLS when given certificate assets
and a private key.

Related issues:
- BLOOM-58 #comment Begin adding mTLS to Java client.
- BLOOM-59 #comment Java client functionality for mTLS.

Worklog:
- [x] Sync schema with `release/1.4.6`
- [ ] Provide constructors/settings for client keys and certs
- [ ] Setup `RPCClient` to work this way
- [ ] Version bump -> `beta11`

Commit:
BLOOM-58: Add client auth support to Java lib

Commit:
BLOOM-58: Add support for client auth in tests and settings

Commit:
Add Makefile command to cut embedded lib

Commit:
BLOOM-59: Expose ClientRPCTest from main library

Commit:
Fix issues reported by Codacy

Commit:
Remove unused imports

sgammon committed:
Merge branch 'master' into release/1.0

sgammon committed:
Merge pull request #22 from Bloombox/fix/BLOOM-58

Feature: mTLS 🙄

Commit:
Update schemas -> tip of master

Commit:
Update schemas

Commit:
Update schema

Commit:
Update order schema

Commit:
Add support for Gradle

Commit:
Add Gradle support to Makefile (i.e. make BUILDMODE=gradle

Commit:
Adapt build to Gradle

Commit:
Base64-encode collection names, as is spec

Commit:
Fix public POM, and by extension, CI build

Commit:
Ignore Order fetch tests for now

Commit:
Disable all tests for order retrieval

Commit:
Disable prod testsuite

Commit:
Use lazy-loaded singletons for each API service

Commit:
Suppress call exceptions in tests (but log them), re-enable tests

Commit:
Fix threading issues in tests

Commit:
Various warning fixes

Commit:
Fix spacing and commenting warnings

Commit:
Additional Kotlin warning fixes

Commit:
Fixup further Kotlin warnings

Commit:
Sync schema

Commit:
Add gradle support to example project

Commit:
[maven-release-plugin] prepare release v1.0-rc1

Commit:
[maven-release-plugin] prepare for next development iteration

Commit:
[maven-release-plugin] prepare release v1.0-rc2

Commit:
[maven-release-plugin] prepare for next development iteration

Commit:
Amend build.gradle to support public build

Commit:
Switch POM files and adjust build flows

Commit:
Add initial CI support for Gradle

Commit:
Build with JDK7 and 9 just for kicks

Commit:
Well, that was a terrible idea.

Commit:
Cleanup and refactor settings construction from Java

Commit:
Update README with new examples and structure

Commit:
Package Kotlin client example in .kotlin, cleanup

Commit:
Add Java client example

Commit:
Define bloombox.version parameter

Commit:
Update module IML

Commit:
[maven-release-plugin] prepare release v1.0-rc7

Commit:
[maven-release-plugin] prepare for next development iteration

Commit:
Testing improvements for menu API

Commit:
Cleanup and improve documentation

Commit:
Issue 1.0-rc8

Commit:
Docs and cleanup in main client

Commit:
Cleanup and bugfixing before release

Commit:
[maven-release-plugin] prepare release v1.0-rc8

Commit:
[maven-release-plugin] prepare for next development iteration

Commit:
Fix proguard

Commit:
Version bump -> 1.0-rc9

Commit:
Fix exception type

Commit:
Fix additional warnings

Commit:
Update Dokka site link

Commit:
Code cleanup and reorgnization before release

Commit:
Add routine to build docs

Commit:
Upgrade Dokka, configure new docs

Commit:
Update module name

Commit:
Further docs improvements

Commit:
Use embedded POM for embedded lib

Commit:
Update site FAQs

Commit:
Add links to examples

Commit:
Tune up/upgrade Maven site content

Commit:
Upgrade Kotlin -> 1.2.21

Commit:
Update Maven plugins

Commit:
Remove reference to latest version

sgammon commented on README.md:
````diff
@@ -45,40 +59,115 @@ If you're developing on the code, you'll use `make`.
+```kotlin
+  // with our client object, obtain shop info synchronously
+  val info = client.shop().info()
+  if (info.getStatus() == ShopStatus.OPEN) print("The shop is OPEN.")
````
lots of docs and examples. you can view these at https://bloombox.github.io/Java/

sgammon commented on build.gradle:
```diff
@@ -0,0 +1,41 @@
+
+tasks.withType(JavaCompile) {
+	options.encoding = 'UTF-8'
+}
```
support for Gradle, so you can `gradle build install` to build/install the library

sgammon commented on example/src/main/java/io/bloombox/example/java/JavaClientExample.java:
```diff
@@ -0,0 +1,87 @@
+/**
+ * Examples for using the Bloombox API client from Java.
+ */
+public final class JavaClientExample {
```
example code in Java

sgammon commented on example/src/main/kotlin/io/bloombox/example/kotlin/KotlinClientExample.kt:
```diff
@@ -34,13 +34,13 @@ object KotlinClientExample {
-  private val client = BloomboxClient(
-        settings = BloomboxClient.Settings(
+  private val client = Bloombox(
+        settings = Bloombox.Settings(
```
example code in Kotlin

Commit:
[maven-release-plugin] prepare release v1.0-rc9

Commit:
[maven-release-plugin] prepare for next development iteration

sgammon reviewed (commented):
additional commentary before release

Commit:
Fix embedded POM

Commit:
Add client lifecycle notes to Dokka site

Commit:
Small docs improvements

Commit:
Remove invalid URL tag from site config

Commit:
Fix sitesearch on docs

samt reviewed (commented):
Link on sidebar (Examples > Java) 404s.

Commit:
Fix forced accesses

Commit:
Fixup complex conditionals

tyler-porras commented on README.md:
```diff
@@ -45,40 +59,115 @@ If you're developing on the code, you'll use `make`.
+- `PICKUP_ONLY`: The storefront is currently open only for pickup orders.
+- `CLOSED`: The storefront is currently closed and not accepting orders of any type.
+
+When an order is submitted to a shop that is `CLOSED` (or a `PICKUP` order is submitted during `DELIVERY_ONLY`, or a
```
Are we planning on implementing functionality so that an order can still be placed/queued during one of these times with a "future fulfillment" that goes through when that order type is accepted again?

sgammon replied:
that's a great point. at the moment orders are just closed depending on the _current_ store status. we should file as a bug and fix it, although that's a server side problem, so it would go in Bloombox/platform.

mstan replied:
I am wondering if accepted future/queued orders would make more sense as a status of its own in this case. Perhaps make it inherently accept future orders while the store is an open state. Otherwise, maybe consider a status where the store isn't open, but is taking queued orders for their next open date. (If the store intends to be closed for a while and not accept future orders, they would just be in the CLOSED state)

sgammon replied:
Good point, and by default at the moment, orders are `PENDING` > `ACCEPTED`/`REJECTED` > (some delivery/pickup specific ones) > `FULFILLED`. We could potentially either add a special status for this case or use `PENDING` to mean that, until they actually fulfill the order

Commit:
Version bump before final release -> 1.0

Commit:
Remove site search

sgammon commented:
fixed sidebar links. they won't work until after the merge but they are correct for the release branch.

samt reviewed (approved):
LGTM :+1:

mstan reviewed (approved):
Curious about future order statuses.  Left comment accordingly. Otherwise, looks ready.

Commit:
Fix doc links

Commit:
Move docs out of pre-release

Commit:
Fix version references -> 1.0

RandyStevens reviewed (approved):
LGTM!

noahdamiani reviewed (approved):
Looks great, going to check out the new java docs page now.
